storage/trace_test.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"testing"
+
+	"cloud.google.com/go/internal/testutil"
+	"cloud.google.com/go/storage/internal"
+	"github.com/google/go-cmp/cmp"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/sdk/trace/tracetest"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/api/option"
+)
+
+type emulatedTraceTest struct {
+	*testing.T
+	resources resources
+}
+
+func TestTraceStorageTraceStartEndSpan(t *testing.T) {
+	ctx := context.Background()
+	e := tracetest.NewInMemoryExporter()
+	tp := sdktrace.NewTracerProvider(sdktrace.WithSyncer(e))
+	defer tp.Shutdown(ctx)
+	otel.SetTracerProvider(tp)
+
+	// TODO: Remove setting development env var upon launch.
+	t.Setenv("GO_STORAGE_DEV_OTEL_TRACING", "true")
+
+	spanName := "storage.TestTrace.TestStorageTraceStartEndSpan"
+	addAttrs := attribute.String("foo", "bar")
+	spanStartOpts := []trace.SpanStartOption{
+		trace.WithAttributes(addAttrs),
+	}
+	newAttrs := attribute.Int("fakeKey", 800)
+
+	ctx, span := startSpan(ctx, spanName, spanStartOpts...)
+	span.SetAttributes(newAttrs)
+	endSpan(ctx, nil)
+
+	spans := e.GetSpans()
+	if len(spans) != 1 {
+		t.Errorf("expected one span, got %d", len(spans))
+	}
+
+	// Test StartSpanOption and Cloud Trace adoption common attributes are appended.
+	// Test startSpan returns the span and additional attributes can be set.
+	wantSpan := createWantSpanStub(spanName)
+	wantSpan.Attributes = append(wantSpan.Attributes, addAttrs)
+	wantSpan.Attributes = append(wantSpan.Attributes, newAttrs)
+	opts := []cmp.Option{
+		cmp.Comparer(spanAttributesComparer),
+	}
+	for _, span := range spans {
+		if diff := testutil.Diff(span, wantSpan, opts...); diff != "" {
+			t.Errorf("diff: -got, +want:\n%s\n", diff)
+		}
+	}
+	e.Reset()
+}
+
+func TestTraceSpansMultiEmulated(t *testing.T) {
+	ctx := skipJSONReads(context.Background(), "no reads in test")
+	// To only test storage library layer instrumentation,
+	// we disable transport layer traces for testing purposes.
+	opts := []option.ClientOption{
+		option.WithTelemetryDisabled(),
+	}
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, bucket string, prefix string, client *Client) {
```

Comment:
Should this be a `transportClientTest` instead?

Reply:
`transportClientTest` has a check to not run if the emulator env var is not set https://github.com/googleapis/google-cloud-go/blob/48022acb5550244ffdff9a87f2792731eec9b005/storage/client_test.go#L2175

We should modify `transportClientTest` to allow configuration of client opts similar to `multiTransportTest`. Or add a check to this test specifically so it only runs on the emulator

Reply:
I believe this caused this issue #11480